keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+```
+// EndpointAddress is a tuple that describes one or more IP addresses for an
+// endpoint.
+type EndpointAddress struct {
+	// The IPs for this endpoint.
+	IPs []string
+	// Optional: Hostname of this endpoint
+	// Meant to be used by DNS servers etc.
+	// +optional
+	Hostname string
+	// Optional: Node hosting this endpoint. This can be used to determine endpoints local to a node.
+	// +optional
+	NodeName *string
+	// Optional: The kubernetes object related to the entry point.
+	TargetRef *ObjectReference
+}
+```
+This internal representation should eventually become part of a versioned API (after a period of deprecation for the singular "IP" field).
+
+#### Maintaining Compatible Interworking between Old and New Clients
+With this API change, we need to consider the scenario where there are a mix of clients that are running old vs. new versions of the API. The old clients would only be aware of the original, singular IP field, whereas newer clients could be writing either just the plural IPs field, or updating both singular IP and plural IPs fields. To cover this case, the API server will need to implement some safeguards/fix-ups for these fields for write operations (in compliance with the guidelines for pluralizing singular API fields that is included in the [Kubernetes API change quidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md):
+
+##### V1 to Core (Internal) Conversion
+- If only V1 IP is provided:
+  - Copy V1 IP to core IPs[0]
+- Else if only V1 IPs[] is provided:
+  - Copy V1 IPs[] to core IPs[]
+- Else if both V1 IP and V1 IPs[] are provided:
+  - Verify that V1 IP matches V1 IPs[0]
+  - Copy V1 IPs[] to core IPs[]
+- Delete any duplicates in core IPs[]
+
+##### Core (Internal) to V1 Conversion
+  - Copy core IPs[0] to V1 IP
+  - Copy core IPs[] to V1 IPs[]
+
+
+#### Configuration of Endpoint IP Family in Service Definitions
+This proposal adds an option to configure an endpoint IP family for a Kubernetes service:
+```
+    endpointFamily: <ipv4|ipv6|dual-stack>       [Default: dual-stack]
+```
+For example, the spec definition for an application that only binds to IPv4 might look like this:
+```
+spec:
+  selector:
+    app: MyApp
+  ports:
+  - protocol: TCP
+    port: 80
+    targetPort: 9376
+  endpointFamily: ipv4
+```
+This per-service endpoint IP family configuration is required for IPv4-only and IPv6-only Application Support. For example, if there is a legacy application that only binds to IPv4 addresses (IPv4-only application) that is running in an otherwise dual-stack cluster, then we don't want endpoints for the service to include IPv6 addresses that are allocated to backend pods for the service. Otherwise, any IPv6 addresses that are included in endpoints would be "dead ends" (no response will be received from the server) for ingress controller and external load balancer operation. The same reasoning applies to IPv6-only applications.
```

Comment:
Either way, this is useful information for a service (and pod ports, while we're at it), because there's nothing stopping a program from binding some things as v4 only and others as v6 only. A bad idea, sure, but technically plausible.

Reply:
I don't want to design to hypotheticals, especially API surface.  If I understand, the worst case is that we publish a v6 nodeport with v6 endpoints and there's nothing listening there.  Is that right?

If so, that doesn't seem like the end of the world to me - I'd rather defer this and come back to it if it is a real problem.